### PR TITLE
Add support for gpt-4o, gpt-4o-mini, text-embedding-3-small and text-embedding-3-large

### PR DESCRIPTION
**Pull request by mbarretol**
This pull request adds support for the new `gpt-4o-mini` model and closes the issue #111, #114 and #115.

### CODE REVIEW

File: js/src/core.ts [hunk 1]
```diff
@@ -275,7 +275,8 @@ export function getEncodingNameForModel(model: TiktokenModel) {
       return "cl100k_base";
     }
     case "gpt-4o":
-    case "gpt-4o-2024-05-13": {
+    case "gpt-4o-2024-05-13":
+    case "gpt-4o-mini": {
```
Comment (Pederzh) [inline js/src/core.ts:279]: We should also add the name of the model version, as is done for other models.
```suggestion
    case "gpt-4o-2024-05-13":
    case "gpt-4o-mini-2024-07-18":
    case "gpt-4o-mini": {
```

File: tiktoken/model_to_encoding.json [hunk 2]
```diff
@@ -54,5 +54,6 @@
     "gpt-4-0125-preview": "cl100k_base",
     "gpt-4-vision-preview": "cl100k_base",
     "gpt-4o": "o200k_base",
-    "gpt-4o-2024-05-13": "o200k_base"
+    "gpt-4o-2024-05-13": "o200k_base",
+    "gpt-4o-mini": "o200k_base"
```
Comment (Pederzh) [inline tiktoken/model_to_encoding.json:58]:
```suggestion
    "gpt-4o-2024-05-13": "o200k_base",
    "gpt-4o-mini-2024-07-18": "o200k_base",
    "gpt-4o-mini": "o200k_base"
```

File: wasm/src/lib.rs [hunk 3]
```diff
@@ -492,6 +493,7 @@ pub fn encoding_for_model(
         "gpt-4-0125-preview" => Ok("cl100k_base"),
         "gpt-4o" => Ok("o200k_base"),
         "gpt-4o-2024-05-13" => Ok("o200k_base"),
+        "gpt-4o-mini" => Ok("o200k_base"),
```
Comment (Pederzh) [inline wasm/src/lib.rs:496]:
```suggestion
        "gpt-4o-2024-05-13" => Ok("o200k_base"),
        "gpt-4o-mini-2024-07-18" => Ok("o200k_base"),
        "gpt-4o-mini" => Ok("o200k_base"),
```

File: wasm/src/lib.rs [hunk 4]
```diff
@@ -420,6 +420,7 @@ export type TiktokenModel =
     | "gpt-4-vision-preview"
     | "gpt-4o"
     | "gpt-4o-2024-05-13"
+    | "gpt-4o-mini"
```
Comment (Pederzh) [inline wasm/src/lib.rs:423]:
```suggestion
    | "gpt-4o-2024-05-13"
    | "gpt-4o-mini-2024-07-18"
    | "gpt-4o-mini"
```

**Review by Pederzh — COMMENTED** [review]
Thanks for adding the new gpt-4o-mini model, since for each model is also handled the "versioned" name, I've added it.

**Comment by danrasmuson** [comment]
Excited for this to merged!

**Comment by theoremonemehdi** [comment]
What is the ETA for this PR to be merged? 
Thanks!

**Comment by Arit143** [comment]
+1 for merging this soon

**Comment by Arit143** [comment]
hey @dqbd , can this be reviewed and merged please? We are quite eager to try out the new gpt4o mini mostly because of the cost efficiency.

**Comment by rynti** [comment]
FYI for anyone eager to just try it out ASAP - as you can see in the implementation of this PR, the underlying token counter seems to be the same as with `gpt-4o`. Therefore you can probably just pass `gpt-4o` as the model name instead of `gpt-4o-mini`.

**Comment by theoremonemehdi** [comment]
@rynti that's exactly why merging this PR is no brainer. I think everybody is already doing something like `!model.startsWith('gpt-4o') ?? 'gpt-4o' `

**Comment by alsotang** [comment]
> @rynti that's exactly why merging this PR is no brainer. I think everybody is already doing something like `model.startsWith('gpt-4o') ? 'gpt-4o' : model`

This method may work currently, but it is not guaranteed. Just like the notorious if condition for determining Windows9x: `if(version.StartsWith("Windows 9"))`

**Comment by huytool157** [comment]
@dqbd could you please take a look and cut a new release if the PR is good?

**Comment by breesetposh** [comment]
bump, can we get this in here? we're going to have to start forking to move this along, but would love to stay on the main release branch

**Review by alexluca89 — APPROVED** [review]

File: js/src/core.ts [hunk 5]
```diff
@@ -275,7 +275,9 @@ export function getEncodingNameForModel(model: TiktokenModel) {
       return "cl100k_base";
     }
     case "gpt-4o":
-    case "gpt-4o-2024-05-13": {
+    case "gpt-4o-2024-05-13":
+    case "gpt-4o-mini-2024-07-18":
+    case "gpt-4o-mini": {
```
Comment (huytool157) [inline js/src/core.ts:280]:
```suggestion
    case "gpt-4o-2024-05-13":
    case "gpt-4o-2024-08-06":
    case "gpt-4o-mini-2024-07-18":
    case "gpt-4o-mini": {
```

File: tiktoken/model_to_encoding.json [hunk 6]
```diff
@@ -54,5 +54,7 @@
     "gpt-4-0125-preview": "cl100k_base",
     "gpt-4-vision-preview": "cl100k_base",
     "gpt-4o": "o200k_base",
-    "gpt-4o-2024-05-13": "o200k_base"
+    "gpt-4o-2024-05-13": "o200k_base",
+    "gpt-4o-mini-2024-07-18": "o200k_base",
+    "gpt-4o-mini": "o200k_base"
```
Comment (huytool157) [inline tiktoken/model_to_encoding.json:59]:
```suggestion
    "gpt-4o-2024-05-13": "o200k_base",
    "gpt-4o-2024-08-06":"o200k_base",
    "gpt-4o-mini-2024-07-18": "o200k_base",
    "gpt-4o-mini": "o200k_base"
```

File: wasm/src/lib.rs [hunk 7]
```diff
@@ -420,6 +420,8 @@ export type TiktokenModel =
     | "gpt-4-vision-preview"
     | "gpt-4o"
     | "gpt-4o-2024-05-13"
+    | "gpt-4o-mini-2024-07-18"
+    | "gpt-4o-mini"
```
Comment (huytool157) [inline wasm/src/lib.rs:424]:
```suggestion
    | "gpt-4o-2024-08-06"
    | "gpt-4o-mini-2024-07-18"
    | "gpt-4o-mini"
```

File: wasm/src/lib.rs [hunk 8]
```diff
@@ -492,6 +494,8 @@ pub fn encoding_for_model(
         "gpt-4-0125-preview" => Ok("cl100k_base"),
         "gpt-4o" => Ok("o200k_base"),
         "gpt-4o-2024-05-13" => Ok("o200k_base"),
+        "gpt-4o-mini-2024-07-18" => Ok("o200k_base"),
+        "gpt-4o-mini" => Ok("o200k_base"),
```
Comment (huytool157) [inline wasm/src/lib.rs:498]:
```suggestion
        "gpt-4o-2024-08-06" => Ok("o200k_base"),
        "gpt-4o-mini-2024-07-18" => Ok("o200k_base"),
        "gpt-4o-mini" => Ok("o200k_base"),
```

**Review by huytool157 — COMMENTED** [review]
Since package author has been unresponsive, I think we should include `gpt-4o-2024-08-06` as well so that we can just release all at once

**Comment by dqbd** [comment]
Hello! Sorry about the long delay! Will take a look as soon as possible!

**Pull request merged**